### ISSUE TITLE
Caching within Matrix: logged in users see public, cached content

### ISSUE THREAD

**Issue by davidjb** [jcu-eresearch/matrix-helpers#20]
On the front-end of our sites, we include certain content and features that are only shown to editors.  For instance, the frontend editing toolbar that appears to guide users to various Matrix links and areas is shown to anyone who is logged in.  However,  there's currently an issue where a logged in user is shown the the 'public' (non-logged in) view of pages if those pages are within the Squiz Edge cache.  A temporary workaround is for the user to bypass the server-side cache with a shift-refresh -- this then forces a bypass of Edge and the page to load fresh from Matrix.  

To reproduce:

1. Log into Matrix at as any user.
2. Do a shift-refresh on a public site and note the blue 'Page tools' menu at the top left; this menu is only shown to logged in users.
3. Open a new Private/Incognito window and visit the same URL -- this loads the page as a public user, placing the page into the Edge Cache.
4. In the first window, do regular refresh (not shift-refresh) of the site
5. Note that the Page Tools menu has disappeared; you are now seeing the Edge cached, public version of this page.

This presents an issue because a user is unable to tell a) if they're logged in or not and b) what the 'true' version of the page is without manually having to shift-refresh -- this is a bigger issue, because it affects the editing process.

As best I can tell, the issue is related to the way the Squiz Edge Content Distribution Network (CDN) is configured.  Testing by toggling Permission-based caching (https://matrix.squiz.net/manuals/system-management/chapters/cache-manager#cache-levels) shows that the Matrix cache (separate to Edge) doesn't appear to be the issue.  Edge is delivering a 'public' (non-logged in) view of the pages to users who are actually logged in -- shown by the `X-Cache: "HIT from squizedge.net"` header coming from the server.  Preferably, what it should be doing is bypassing its cache for users who are logged in (eg testing for presence of a login cookie etc), thus ensuring an editor or admin is always seeing the 'correct' and latest page.  

Related issues: 
* https://squizmap.squiz.net/matrix/9846
* https://forums.squiz.net/t/browser-cache-control-for-logged-in-users/8134
* https://forums.squiz.net/t/matrix-cache-issue/5538
* https://forums.squiz.net/t/matrix-cache-issue/5538/7 (particularly where Greg of Squiz highlights that the issue isn't Matrix but is the cache in front of Matrix; eg Edge)
* https://forums.squiz.net/t/logging-in-and-caching/4048

**Comment by davidjb** [comment]
I found the URL `/_noproxycache` which can be used to show the uncached, true version of any given page.  This proves that it's the fault of Squiz Edge.

The good news is that it's possible to create a hack to solve this:  
https://github.com/jcu-eresearch/matrix-helpers/blob/master/web-framework/designs/landing.html#L40
https://github.com/jcu-eresearch/matrix-helpers/blob/master/web-framework/designs/landing.html#L367

Which is essentially redirecting a client in JS if they have been seen to have been logged in (because you can't actually tell client side whether someone is logged in...yet) to the cache bypass URL and then rewriting all the rest of the links in the page to being `/_nocacheproxy`.  Works for now :)

Reported to Squiz support, reference number is: 117992

**Comment by davidjb** [comment]
Matrix 5.4 added the secret script `/__lib/edge/authsession.php` which produces JSON data which tells us `user_is_public: "[0 or 1]"`.  Apparently this can be used by Edge to cache pages based upon user permissions, but what we actually want is pages to *not be cached at all* when a user is logged in because we show all sorts of data specific to them (username, lists of content only they might have access to etc etc).

**Comment by davidjb** [comment]
Had a discussion with Squiz about this.  The tl;dr is that they won't/can't fix it.

Edge is incapable of bypassing its cache for anything but *all* users in the system and Squiz staunchly refuse to do this because of the apparently performance detriment it will cause.  A slew of reasons as to why `authsession.php` wasn't designed to work with Edge followed, together with an up-sell to Trinity, a brand new cache product from Squiz that would *apparently* do what we need.

Getting to the technical details, the current Edge system can only be bypassed by a) `Cache-Control` headers or b) `/_noproxycache` or `/_nocache` URLs.  Since we can't get a browser to send `Cache-Control` with every request, we're left with relying on URL hacking if a user is logged in. 

In short, what we've done is actually the only option that's possible, so well done on doing the very best we can.  Closing as `wontfix` as the vendor won't provide a solution and unfortunately doesn't quite understand that editors actually want to try their sites / changes before going live.
